File: src/utils/calculations.ts
import {
  type PersonalInfo,
  type SalaryCalculation,
  type BudgetSummary,
  type ExpenseItem,
  type SalaryInfo,
} from "../types";
import { getApplicableTable, SOCIAL_SECURITY_RATE } from "../data/irsTables";

/**
 * Calcula a retenção de IRS para um dado salário bruto
 *
 * Fórmula: Retenção = Remuneração × Taxa - Parcela_a_abater - (Parcela_dependente × N_dependentes)
 */
export function calculateIRSRetention(
  grossSalary: number,
  personalInfo: PersonalInfo,
): { retention: number; rate: number } {
  if (grossSalary <= 0) return { retention: 0, rate: 0 };

  const table = getApplicableTable(
    personalInfo.maritalStatus,
    personalInfo.titulares,
    personalInfo.dependentes,
  );

  const bracket = table.brackets.find((b) => grossSalary <= b.upTo);
  if (!bracket || bracket.rate === 0) {
    return { retention: 0, rate: 0 };
  }

  const retention =
    grossSalary * bracket.rate -
    bracket.parcelaAbater -
    bracket.parcelaDependente * personalInfo.dependentes;

  // A retenção nunca pode ser negativa
  const finalRetention = Math.max(0, Math.round(retention * 100) / 100);
  const effectiveRate = grossSalary > 0 ? finalRetention / grossSalary : 0;

  return {
    retention: finalRetention,
    rate: Math.round(effectiveRate * 10000) / 10000,
  };
}

/**
 * Calcula a contribuição para a Segurança Social
 * Taxa fixa de 11% para trabalhadores por conta de outrem
 */
export function calculateSocialSecurity(grossSalary: number): number {
  if (grossSalary <= 0) return 0;
  return Math.round(grossSalary * SOCIAL_SECURITY_RATE * 100) / 100;
}

/**
 * Calcula o salário líquido completo
 */
export function calculateNetSalary(
  grossSalary: number,
  personalInfo: PersonalInfo,
): SalaryCalculation {
  if (grossSalary <= 0) {
    return {
      grossAmount: 0,
      irsRetention: 0,
      irsRate: 0,
      socialSecurity: 0,
      socialSecurityRate: SOCIAL_SECURITY_RATE,
      netAmount: 0,
    };
  }

  const { retention, rate } = calculateIRSRetention(grossSalary, personalInfo);
  const ss = calculateSocialSecurity(grossSalary);
  const net = Math.round((grossSalary - retention - ss) * 100) / 100;

  return {
    grossAmount: grossSalary,
    irsRetention: retention,
    irsRate: rate,
    socialSecurity: ss,
    socialSecurityRate: SOCIAL_SECURITY_RATE,
    netAmount: Math.max(0, net),
  };
}

/**
 * Calcula o resumo completo do orçamento mensal
 */
export function calculateBudgetSummary(
  salaries: [SalaryInfo, SalaryInfo],
  personalInfo: PersonalInfo,
  expenses: ExpenseItem[],
): BudgetSummary {
  const salary1 = salaries[0].enabled
    ? calculateNetSalary(salaries[0].grossAmount, personalInfo)
    : calculateNetSalary(0, personalInfo);

  const salary2 = salaries[1].enabled
    ? calculateNetSalary(salaries[1].grossAmount, personalInfo)
    : calculateNetSalary(0, personalInfo);

  const totalGross = salary1.grossAmount + salary2.grossAmount;
  const totalNet = salary1.netAmount + salary2.netAmount;
  const totalIRS = salary1.irsRetention + salary2.irsRetention;
  const totalSS = salary1.socialSecurity + salary2.socialSecurity;
  const totalDeductions = totalIRS + totalSS;

  const totalExpenses = expenses
    .filter((e) => e.enabled)
    .reduce((sum, e) => sum + e.amount, 0);

  const netLiquidity = Math.round((totalNet - totalExpenses) * 100) / 100;
  const savingsRate = totalNet > 0 ? netLiquidity / totalNet : 0;

  return {
    salary1,
    salary2,
    totalGross,
    totalNet,
    totalIRS,
    totalSS,
    totalDeductions,
    totalExpenses,
    netLiquidity,
    savingsRate: Math.round(savingsRate * 10000) / 10000,
  };
}

/**
 * Formata um valor em euros
 */
export function formatCurrency(value: number): string {
  return new Intl.NumberFormat("pt-PT", {
    style: "currency",
    currency: "EUR",
  }).format(value);
}

/**
 * Formata uma percentagem
 */
export function formatPercentage(value: number): string {
  return new Intl.NumberFormat("pt-PT", {
    style: "percent",
    minimumFractionDigits: 1,
    maximumFractionDigits: 1,
  }).format(value);
}

File: monthy_budget_flutter/lib/utils/calculations.dart
import 'dart:math' as math;
import '../models/app_settings.dart';
import '../models/budget_summary.dart';
import '../data/tax/tax_system.dart';
import '../data/irs_tables.dart';

double _round2(double v) => (v * 100).roundToDouble() / 100;

/// Calcula o subsidio de alimentação mensal (PT only)
MealAllowanceCalculation calculateMealAllowance(
  MealAllowanceType type,
  double perDay,
  int workingDays,
  double irsRate,
  double ssRate,
) {
  if (type == MealAllowanceType.none || perDay <= 0 || workingDays <= 0) {
    return const MealAllowanceCalculation();
  }

  final exemptLimit = type == MealAllowanceType.card ? mealCardExemptLimit : mealCashExemptLimit;
  final totalMonthly = _round2(perDay * workingDays);
  final exemptPerDay = math.min(perDay, exemptLimit);
  final taxablePerDay = math.max(0.0, perDay - exemptLimit);

  final exemptPortion = _round2(exemptPerDay * workingDays);
  final taxablePortion = _round2(taxablePerDay * workingDays);

  final irsTaxOnMeal = _round2(taxablePortion * irsRate);
  final ssTaxOnMeal = _round2(taxablePortion * ssRate);
  final netMealAllowance = _round2(totalMonthly - irsTaxOnMeal - ssTaxOnMeal);

  return MealAllowanceCalculation(
    totalMonthly: totalMonthly,
    exemptPortion: exemptPortion,
    taxablePortion: taxablePortion,
    irsTaxOnMeal: irsTaxOnMeal,
    ssTaxOnMeal: ssTaxOnMeal,
    netMealAllowance: netMealAllowance,
  );
}

/// Calcula o salário líquido completo, usando o sistema fiscal do país selecionado
SalaryCalculation calculateNetSalary(
  SalaryInfo salary,
  PersonalInfo personalInfo,
  TaxSystem taxSystem,
) {
  final country = taxSystem.country;
  final baseGross = salary.grossAmount;

  // Subsidies: PT and ES have duodécimos/pagas extras, FR and UK don't
  final factor = country.hasSubsidies ? salary.subsidyMode.monthlyFactor : 1.0;
  final effectiveGross = _round2(baseGross * factor);
  final subsidyBonus = _round2(effectiveGross - baseGross);

  if (effectiveGross <= 0) {
    final meal = country.hasMealAllowance
        ? calculateMealAllowance(
            salary.mealAllowanceType,
            salary.mealAllowancePerDay,
            salary.workingDaysPerMonth,
            0,
            taxSystem.socialContributionRate,
          )
        : const MealAllowanceCalculation();
    return SalaryCalculation(
      grossAmount: baseGross,
      effectiveGrossAmount: 0,
      subsidyMonthlyBonus: 0,
      otherExemptIncome: salary.otherExemptIncome,
      irsRetention: 0,
      irsRate: 0,
      socialSecurity: 0,
      socialSecurityRate: taxSystem.socialContributionRate,
      netAmount: 0,
      mealAllowance: meal,
      totalNetWithMeal: _round2(meal.netMealAllowance + salary.otherExemptIncome),
    );
  }

  final taxResult = taxSystem.calculateTax(
    grossSalary: effectiveGross,
    maritalStatus: personalInfo.maritalStatus.jsonValue,
    titulares: salary.titulares,
    dependentes: personalInfo.dependentes,
  );

  final netAmount = math.max(0.0, _round2(effectiveGross - taxResult.incomeTax - taxResult.socialContribution));

  final meal = country.hasMealAllowance
      ? calculateMealAllowance(
          salary.mealAllowanceType,
          salary.mealAllowancePerDay,
          salary.workingDaysPerMonth,
          taxResult.incomeTaxRate,
          taxSystem.socialContributionRate,
        )
      : const MealAllowanceCalculation();

  return SalaryCalculation(
    grossAmount: baseGross,
    effectiveGrossAmount: effectiveGross,
    subsidyMonthlyBonus: subsidyBonus,
    otherExemptIncome: salary.otherExemptIncome,
    irsRetention: taxResult.incomeTax,
    irsRate: taxResult.incomeTaxRate,
    socialSecurity: taxResult.socialContribution,
    socialSecurityRate: taxSystem.socialContributionRate,
    netAmount: netAmount,
    mealAllowance: meal,
    totalNetWithMeal: _round2(netAmount + meal.netMealAllowance + salary.otherExemptIncome),
  );
}

/// Calcula o resumo completo do orçamento mensal
BudgetSummary calculateBudgetSummary(
  List<SalaryInfo> salaries,
  PersonalInfo personalInfo,
  List<ExpenseItem> expenses,
  TaxSystem taxSystem, {
  Map<String, double> monthlyBudgets = const {},
}) {
  final calcs = salaries.map((s) {
    if (!s.enabled) return const SalaryCalculation();
    return calculateNetSalary(s, personalInfo, taxSystem);
  }).toList();

  final totalGross = calcs.fold(0.0, (sum, s) => sum + s.effectiveGrossAmount);
  final totalNet = calcs.fold(0.0, (sum, s) => sum + s.netAmount);
  final totalMealAllowance = calcs.fold(0.0, (sum, s) => sum + s.mealAllowance.netMealAllowance);
  final totalNetWithMeal = calcs.fold(0.0, (sum, s) => sum + s.totalNetWithMeal);
  final totalIRS = calcs.fold(0.0, (sum, s) => sum + s.irsRetention + s.mealAllowance.irsTaxOnMeal);
  final totalSS = calcs.fold(0.0, (sum, s) => sum + s.socialSecurity + s.mealAllowance.ssTaxOnMeal);
  final totalDeductions = totalIRS + totalSS;

  // Sum per-category default amounts, then apply monthly overrides
  final defaultByCategory = <String, double>{};
  for (final e in expenses.where((e) => e.enabled)) {
    defaultByCategory[e.category.name] =
        (defaultByCategory[e.category.name] ?? 0) + e.amount;
  }
  final totalExpenses = defaultByCategory.entries.fold(0.0, (sum, entry) {
    return sum + (monthlyBudgets[entry.key] ?? entry.value);
  });

  final netLiquidity = _round2(totalNetWithMeal - totalExpenses);
  final savingsRate = totalNetWithMeal > 0 ? netLiquidity / totalNetWithMeal : 0.0;

  return BudgetSummary(
    salaries: calcs,
    totalGross: totalGross,
    totalNet: totalNet,
    totalNetWithMeal: totalNetWithMeal,
    totalMealAllowance: totalMealAllowance,
    totalIRS: totalIRS,
    totalSS: totalSS,
    totalDeductions: totalDeductions,
    totalExpenses: totalExpenses,
    netLiquidity: netLiquidity,
    savingsRate: (savingsRate * 10000).roundToDouble() / 10000,
  );
}
